### PR TITLE
[REVIEW] Remove bounds check for `cudf::gather`

### LINKED ISSUE #6478: [FEA] Remove bounds checking from `cudf::gather`
This is the first of a series of issues as part of a larger `join` refactor.

Currently, `cudf::gather` accepts a `check_bounds` parameter. We should remove that and move the bounds checking logic to Cython, as detailed in https://github.com/rapidsai/cudf/issues/5505.

### LINKED ISSUE #6731: [FEA] `minmax` reduction binding in Cython/Python
**Is your feature request related to a problem? Please describe.**
Currently libcudf supports `minmax` reduction #6247 , but cython binding is lacking.

**Describe the solution you'd like**
Create cython binding for `cudf::minmax`

### PR DESCRIPTION
**Pull request by isVoid**
Closes #6478 

`cudf::gather` will not run a pre-pass to check for index validity.

For `out_of_bounds_policy`, remove `FAIL`, and expose `NULLIFY` and `DONT_CHECK` to user. `NULLIFY` sets out of bounds indices to null rows, while `DONT_CHECK` skips any checking. Using `DONT_CHECK` should yield higher performance to gather maps with only valid indices.

Note that the negative index (wrap-arounds) policy is unchanged. When gather map dtype is signed, wrap-around is applied.

A new Cython binding to `cudf::minmax`, used for Cython bound checking is added. Will also closes #6731

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/cudf/copying.hpp
Comment (jrhemstad) [inline cpp/include/cudf/copying.hpp:76]: What we really want is to control what happens to negative indices. Do they wrap around or do they count as out of bounds?

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/copying.hpp
Comment (shwina) [reply cpp/include/cudf/copying.hpp:76]: Wrap around (opt-in)

Comment (jrhemstad) [reply cpp/include/cudf/copying.hpp:76]: Is the plan to add an additional parameter to control that behavior?

**Comment by ajschmidt8** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

File: cpp/include/cudf/copying.hpp
Comment (isVoid) [reply cpp/include/cudf/copying.hpp:76]: We haven't discussed about it yet. Personally, I think we should remove the support of wrap-around for out-of-bound negative indices and only support for range `[-n_rows, n_rows)`. Thoughts?

Comment (jrhemstad) [reply cpp/include/cudf/copying.hpp:76]: The previous logic makes sense to me. If the gather map is unsigned, then there aren't any negative values to wrap around.

Comment (shwina) [reply cpp/include/cudf/copying.hpp:76]: I'm OK with not exposing a parameter and implicitly using the signed-ness of the gathermap too. It doesn't matter for Python since we're _always_ going to need negative index handling.

File: cpp/include/cudf/copying.hpp
Comment (jrhemstad) [inline cpp/include/cudf/copying.hpp]:
```suggestion
 * better performance. Behavior is undefined if `DONT_CHECK` is used and the gather map contains out-of-bound indices. 
```

File: cpp/include/cudf/copying.hpp
Comment (jrhemstad) [inline cpp/include/cudf/copying.hpp:43]: Move the docs from above. 
```suggestion
enum class out_of_bounds_policy : int8_t { 
   NULLIFY, /// Output values corresponding to out-of-bounds indices are null
   DONT_CHECK /// No bounds checking is performed
};
```

**Comment by isVoid** [comment]
rerun tests

**Comment by isVoid** [comment]
rerun tests

File: python/cudf/cudf/_lib/copying.pyx
Comment (shwina) [inline python/cudf/cudf/_lib/copying.pyx:155]: Can't wait until we have `black` for Cython so that we don't have to debate style, but I'd prefer:

```suggestion
    cdef cpp_copying.out_of_bounds_policy policy = (
        cpp_copying.out_of_bounds_policy.DONT_CHECK
    )
```

The current style seems more suited for method chaining.

File: python/cudf/cudf/_lib/copying.pyx
Comment (shwina) [inline python/cudf/cudf/_lib/copying.pyx:143]: Users don't know what a `Table` is:

```suggestion
                             f" {gm_max} is out of bound in {type(source_table)} with"
```

File: cpp/include/cudf/copying.hpp
Comment (jrhemstad) [inline cpp/include/cudf/copying.hpp:81]: This should default to `DONT_CHECK` to more closely match the original behavior.

File: cpp/src/dictionary/decode.cu
Comment (jrhemstad) [inline cpp/src/dictionary/decode.cu:47]: This seems like it would change the original behavior.

File: cpp/src/dictionary/remove_keys.cu
Comment (jrhemstad) [inline cpp/src/dictionary/remove_keys.cu:115]: This too.

File: cpp/src/groupby/sort/group_argmax.cu
Comment (jrhemstad) [inline cpp/src/groupby/sort/group_argmax.cu:54]: This too.

File: cpp/src/groupby/sort/group_argmin.cu
Comment (jrhemstad) [inline cpp/src/groupby/sort/group_argmin.cu:54]: And this.

File: cpp/src/groupby/sort/groupby.cu
Comment (jrhemstad) [inline cpp/src/groupby/sort/groupby.cu:204]: And this.

File: cpp/src/groupby/sort/groupby.cu
Comment (jrhemstad) [inline cpp/src/groupby/sort/groupby.cu:241]: And this.

File: cpp/src/transform/encode.cu
Comment (jrhemstad) [inline cpp/src/transform/encode.cu:79]: This too.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
It seems like there are a lot of places where the original behavior has been changed from `DONT_CHECK` to `NULLIFY`.

**Comment by jrhemstad** [comment]
Can you also update the `detail::gather` to use `out_of_bounds_policy` instead of the raw bool here: https://github.com/rapidsai/cudf/blob/branch-0.17/cpp/include/cudf/detail/gather.cuh#L622

**Comment by isVoid** [comment]
> It seems like there are a lot of places where the original behavior has been changed from `DONT_CHECK` to `NULLIFY`.

I think the replacements result from a change to this line: https://github.com/rapidsai/cudf/blob/fb789bfa51a4fdab9f02fd9c509eaaca8de511da/cpp/src/copying/gather.cu#L70
Currently when calling `cudf::gather` with `NULLIFY` actually gets redirected to `IGNORE` code path, and this PR includes a fix to it. Accordingly, all calls to `gather` should be flipped to maintain code semantics (currently only partially replaced).

Addressing your second review comment, this line will be updated without using booleans.

File: cpp/src/transform/encode.cu
Comment (isVoid) [reply cpp/src/transform/encode.cu:79]: `FAIL` policy is no longer supported, I believe I checked this with Ashwin that the replacement is fine.

Could you confirm? @shwina

Comment (jrhemstad) [reply cpp/src/transform/encode.cu:79]: I think this should be `IGNORE` then.

**Review by harrism — CHANGES_REQUESTED** [review]
Just suggesting doc improvements. Note "bounds" is plural in "out of bounds" and "bounds checking".

File: cpp/src/transform/encode.cu
Comment (isVoid) [reply cpp/src/transform/encode.cu]: Done.

**Comment by isVoid** [comment]
Offline discussion with @jrhemstad : Instead of changing `expect_equal` to `expect_equivalent` and force the test cases pass, let's  investigate each `gather` API call and make sure the semantics remain unchanged before and after the refactor.

**Review by harrism — APPROVED** [review]

File: cpp/src/dictionary/add_keys.cu
Comment (isVoid) [inline cpp/src/dictionary/add_keys.cu:93]: @davidwendt can you take a look if this is correct?

File: cpp/src/dictionary/decode.cu
Comment (isVoid) [reply cpp/src/dictionary/decode.cu:50]: @davidwendt and this?

**Comment by isVoid** [comment]
@devavret can you take a look at `groupby` related code of this refactor when you have a chance?

File: python/cudf/cudf/_lib/reduce.pyx
Comment (shwina) [inline python/cudf/cudf/_lib/reduce.pyx:117]: Needs a `Returns` section.

Side note/opinion: these are internal functions so we don't really need to document them in full numpydoc style. Just a one-liner explaining what the function returns is sufficient. Documenting internals in so much detail can lead to a lot of unnecessary churn.

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

**Review by devavret — APPROVED** [review]
Reviewing groupby changes, I had the same concerns as Jake about change of behaviour. Comparing all the changes with my original version (when nullify out of bounds was a bool), they appear correct. @isVoid is right, there seems to be a bug introduced in gather when the bool parameters were changed to enum. The roles of NULLIFY and IGNORE were reversed.

Approving the groupby changes in this PR.

**Review by nvdbaranec — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by isVoid** [comment]
Rebasing to `branch-0.17`

**Comment by isVoid** [comment]
Closing and defer to another PR for rebase.

**Pull request closed without merging**
